src/clj/game/cards-upgrades.clj

```diff
@@ -175,6 +175,33 @@
                                        (move state side c :hand)
                                        (update-run-ice state side)))} card nil)))}]}
 
+   "Mumbad City Grid"
+   {:abilities [{:req (req (and this-server
+                                (< (:position run) (count (:ices run)))
+                                (> (count (:ices run)) 1)))
+                 :label "Swap the ICE just passed with another piece of ICE protecting this server"
+                 :effect (req (let [passed-ice (nth (get-in @state (vec (concat [:corp :servers] (:server run) [:ices])))
+                                                                                (:position run))
+                                    ice-zone (:zone passed-ice)]
+                                 (resolve-ability state :corp
+                                   {:prompt (msg "Select a piece of ICE to swap with " (:title passed-ice))
+                                    :choices {:req #(and (= ice-zone (:zone %))
+                                                         (is-type? % "ICE"))}
+                                    :effect (req (let [fndx (ice-index state passed-ice)
+                                                       sndx (ice-index state target)
+                                                       fnew (assoc passed-ice :zone (:zone target))
+                                                       snew (assoc target :zone (:zone passed-ice))]
+                                                   (swap! state update-in (cons :corp ice-zone)
+                                                          #(assoc % fndx snew))
+                                                   (swap! state update-in (cons :corp ice-zone)
+                                                          #(assoc % sndx fnew))
+                                                   (update-ice-strength state side fnew)
+                                                   (update-ice-strength state side snew)))} card nil)
+                                 (system-msg state side (str "uses Mumbad City Grid to swap "
+                                                             (if (:rezzed passed-ice) (:title passed-ice) "a piece of ICE")
+                                                             " with "
+                                                             (if (:rezzed target) (:title target) "a piece of ICE")))))}]}
```

Comment:
I think I know why that's the case - the message was printed to the chatlog before even prompt selecting the ICE was finished. So the `target` could have just been `nil` at that moment...